Solutions/Workshops/workshop4.lean

```diff
@@ -0,0 +1,331 @@
+  rcases xtu with xt | xu
+  · left
+    show x ∈ s ∧ x ∈ t
+-- `show` is another tactic for human readability (Lean does not need it)
+-- We declare the type of the goal we are going to prove.
+-- It can be used to rewrite a goal to something definitionally equal.
+    exact ⟨xs, xt⟩
+  . right
+    show x ∈ s ∧ x ∈ u
+    exact ⟨xs, xu⟩
+  tada
+
+example : (s \ t) \ u ⊆ s \ (t ∪ u) := by
+  intro x xstu
+  have xs : x ∈ s := xstu.1.1   -- `x ∈ s \ t` means `x ∈ s ∧ x ∉ t`
+  have xnt : x ∉ t := xstu.1.2
+  have xnu : x ∉ u := xstu.2
+  constructor
+  · exact xs
+  intro xtu
+  -- `x ∈ t ∨ x ∈ u`
+  rcases xtu with xt | xu
+  · exact xnt xt
+  . exact xnu xu
+  tada
+
+-- This is quite lengthy. We can use `rintro` to shorten the proof
+
+example : (s \ t) \ u ⊆ s \ (t ∪ u) := by
+  rintro x ⟨⟨xs, xnt⟩, xnu⟩
+  use xs
+  rintro (xt | xu) <;> contradiction -- We often need brackets when using `rintro` on a conjunction
+  tada
+
+-- Two sets are equal when they have the same elements. This is called "extensionality"
+-- You can use it by typing `ext _`
+-- This is the same tactic used to prove functions are equal if they agree on all inputs
+-- When we define a structure with extensionality-like property, we can mark it with `@[ext]`
+-- so that Lean adds an appropriate extensionality lemma to the `ext` tactic.
+-- More on this in the next lectures.
+
+
+example : s ∩ t = t ∩ s := by
+  ext x -- You can give the variable a name by typing it instead of `_`
+  constructor
+  · rintro ⟨xs, xt⟩; exact ⟨xt, xs⟩
+  . rintro ⟨xt, xs⟩; exact ⟨xs, xt⟩
+  tada
+
+-- A similar statement is the theorem `Subset.antisymm`
+
+#check Subset.antisymm
+
+-- Membership to intersection and union can be unfolded using the following two theorems
+-- (although Lean usually does not need it)
+
+#check mem_union
+#check mem_inter_iff
+
+-- Exercises (MIL 4.1)
+
+-- 1
+example : s ∩ (s ∪ t) = s := by
+  ext x
+  constructor
+  . rintro ⟨xs, _⟩
+    exact xs
+  . rintro xs
+    exact ⟨xs, Or.inl xs⟩
+  tada
+-- another way:
+example : s ∩ (s ∪ t) = s := by
+  ext x
+  simp only [mem_inter_iff, mem_union, and_iff_left_iff_imp]
+  exact λ xs ↦ Or.inl xs
+
+
+-- 2
+example : s ∪ s ∩ t = s := by
+  ext x
+  constructor
+  . rintro (xs | ⟨xs, _⟩)
+    exact xs
+    exact xs
+    tada
+  . intro xs
+    exact Or.inl xs
+    tada
+  tada
+
+-- 3
+example : s \ t ∪ t = s ∪ t := by
+  ext x
+  constructor
+  . rintro (xs | xt)
+    . simp only [mem_union]
+      exact Xor'.or (Or.inl xs)
+    . right
+      exact xt
+  . rintro (xs | xt)
+    . simp only [mem_union, mem_diff]
+      by_cases h : x ∈ t
+      . right
+        exact h
+      . left
+        exact ⟨xs, h⟩
+    . right
+      exact xt
+  tada
+
+-- 4
+example : s \ t ∪ t \ s = (s ∪ t) \ (s ∩ t) := by
+  ext x
+  simp only [mem_union, mem_diff, mem_inter_iff]
+  constructor
+  . rintro (⟨xs, xnt⟩ | ⟨xs, xnt⟩) <;>
+    simp only [xnt, and_false, not_false_iff, and_true, or_false, xs, false_or]
+  .
```

Comment:
Why do we have a new line here? Instead of:
```suggestion
  . rintro ⟨(xts | xts), xs⟩ <;>
```

Reply:
I think it was for readability. Because it extended to more than 100 spaces in the column.

You can change it to be in one line though, if you think it might confuse them. Unless this would be a good time to say that you can have a new line, just as long as it doesn't screw up the indentation?